hilti/toolchain/src/compiler/optimizer/optimizer.cc

```diff
@@ -0,0 +1,434 @@
+    // left unset during insertion into the set).
+    for ( auto& [function, module] : _modified_functions ) {
+        assert(! module); // was left unset
+
+        module = function->parent<declaration::Module>();
+        if ( ! module )
+            // This can happen if the function has been removed from the AST in
+            // the meantime. We still leave the function in the set, but let
+            // its module remain unset.
+            HILTI_DEBUG(logging::debug::Optimizer,
+                        util::fmt("  - skipping function %s as its no longer part of the AST", function->id()));
+    }
+
+    // Go through cached CFGs and invalidate any whose block is no longer part
+    // of the AST.
+    for ( auto& [block, cfg] : _cfgs ) {
+        if ( ! block->parent<declaration::Module>() )
+            cfg = nullptr;
+    }
+}
+
+void ASTState::updateAST(const PassInfo& pinfo) {
+    util::timing::Collector _1("hilti/compiler/optimizer/update-state");
+    util::timing::Collector _2(util::fmt("hilti/compiler/optimizer/update-state/%s", pinfo.name));
+
+    auto run_on_changed_nodes = [&](std::string_view post_processor, const auto& callback) -> bool {
+        util::timing::Collector _(util::fmt("hilti/compiler/optimizer/update-state/%s/%s", pinfo.name, post_processor));
+
+        bool modified = false;
+
+        logging::DebugPushIndent _1(logging::debug::Optimizer);
+        HILTI_DEBUG(logging::debug::Optimizer, util::fmt("* %s", post_processor));
+        logging::DebugPushIndent _2(logging::debug::Optimizer);
+
+        for ( const auto& [function, module] : _modified_functions ) {
+            if ( ! module )
+                continue; // no longer in AST
+
+            if ( _modified_modules.contains(module) ) // will be handled when processing module
+                continue;
+
+            HILTI_DEBUG(logging::debug::Optimizer, util::fmt("- updating function: %s", function->id()));
+            modified |= callback(function->parent());
+        }
+
+        for ( const auto& m : _modified_modules ) {
+            HILTI_DEBUG(logging::debug::Optimizer, util::fmt("- updating module: %s", m->id()));
+            modified |= callback(m);
+        }
+
+        return modified;
+    };
+
+    if ( pinfo.guarantees == bitmask<Guarantees>(Guarantees::All) )
+        return;
+
+    HILTI_DEBUG(logging::debug::Optimizer,
+                util::fmt("re-resolving AST assuming guarantees %s", to_string(pinfo.guarantees)));
+
+    _normalizeModificationState();
+
+    unsigned int round = 1;
+
+    while ( true ) {
+        // The following mimics ASTContext::_resolve(), but skips steps that
+        // the pass doesn't require. It also runs only on changed parts of the
+        // AST; it needs to run only HILTI's versions, no
+        // need to consider other compiler plugins.
+        HILTI_DEBUG(logging::debug::Optimizer, util::fmt("re-resolving AST, round %d", round));
```

Comment:
"re-resolving" feels a bit too narrow here; perhaps "re-processing"?

Also just a question: why no need to consider compiler plugins here?

Reply:
Because everything has already been lowered to HILTI at the time the optimizer runs, no need to re-process any, e.g., Spicy-side elements. I'll add that to the comment.